hip/CMakeLists.txt

```diff
@@ -110,14 +84,14 @@ foreach(GKO_JACOBI_BLOCK_SIZE IN LISTS GKO_HIP_JACOBI_BLOCK_SIZES)
         ${CMAKE_CURRENT_BINARY_DIR}/preconditioner/jacobi_advanced_apply_kernels.instantiate.${GKO_JACOBI_BLOCK_SIZE}.hip.cpp)
 endforeach()
 string(REPLACE ";" "," GKO_HIP_JACOBI_BLOCK_SIZES_CODE "${GKO_HIP_JACOBI_BLOCK_SIZES}")
-configure_file(preconditioner/jacobi_common.hip.hpp.in preconditioner/jacobi_common.hip.hpp)
+configure_file(preconditioner/jacobi_common.hip.hpp.in common/cuda_hip/preconditioner/jacobi_common.hpp)
```

Comment:
the file from cuda and hip will overwrite each other when enable both.

Reply:
no, this happens inside the `cuda` and `hip` subdirectory of your build directory. I tested the builds with a CUDA + ROCm environment the entire time.

Reply:
ah, okay. it is in the <ginkgo>/cuda/common/cuda_hip/preconditioner.hpp and <ginkgo>/hip/common/cuda_hip/preconditioner.hpp.
My concern is not solved by this.
I was thinking both are in <ginkgo>/common/cuda_hip/...

Reply:
It is not clear to me what the actual concern is - can you please elaborate?

All solutions I've seen proposed so far are much more complicated - this solution preserves the directory structure and chooses a different file for each backend, addressing both concerns you mentioned previously. `#include` statements can't appear inside a macro, so the `CMAKE_DISPATCH` suggestion won't work.

Reply:
I think you try to unify the include path of the header by adding the current folder not the root/top folder as the search path.
I mean something like the `#include CMAKE_GENERATED_HEADER(real source file)`, but it is not necessary anyways.
The original one does not change the directory structure and is not complicated. 
I am confused about the location and it sounds unnecessary to me because those files using that are not unified.
I would like avoid exception if it is not necessary. (I call it exception just from the ginkgo experience)

Reply:
This only concerns a single header file (and it doesn't look like we will add any more like this in the near future), so I think we should not start building anything complex around it. `jacobi_kernels.cpp` is already unified, and the other Jacobi files will be unified in a subsequent PR (the whole block size instantiation setup requires some more work that I wanted to keep separate).

Reply:
I see. I was missing the unified jacobi_kernels.cpp one.
Then, preconditioner/jacobi_common.hpp is okay to me with the requirement from the unified one.